### ISSUE TITLE
SIGILL with method definition in top-level block

### ISSUE THREAD

**Issue by jw3126** [JuliaLang/julia#24316]
The following code segfaults on nightly (gives an error on 0.6):
```julia
using Test

f() = false

@testset "error message call" begin
    f(m) = false
    if !(f(1))
        $QuoteNode(:(f(2)))
    end
end
```

**Comment by yuyichao** [comment]
I'm getting an unreachable and not segfault.

**Comment by jw3126** [comment]
Sorry you are right, the full error is
```
Unreachable reached at 0x7fa9a5e96b75

signal (4): Illegal instruction
in expression starting at /home/jan/.julia/v0.7/ArgCheck/test/mvp.jl:10
macro expansion at /home/jan/.julia/v0.7/ArgCheck/test/mvp.jl:13 [inlined]
macro expansion at /home/jan/juliadev/usr/share/julia/site/v0.7/Test/src/Test.jl:950 [inlined]
anonymous at ./<missing> (unknown line)
jl_call_fptr_internal at /home/jan/juliadev/src/julia_internal.h:366 [inlined]
jl_call_method_internal at /home/jan/juliadev/src/julia_internal.h:385 [inlined]
jl_toplevel_eval_flex at /home/jan/juliadev/src/toplevel.c:652
jl_parse_eval_all at /home/jan/juliadev/src/ast.c:798
jl_load at /home/jan/juliadev/src/toplevel.c:687
include_relative at ./loading.jl:533
jl_call_fptr_internal at /home/jan/juliadev/src/julia_internal.h:366 [inlined]
jl_call_method_internal at /home/jan/juliadev/src/julia_internal.h:385 [inlined]
jl_apply_generic at /home/jan/juliadev/src/gf.c:2003
include at ./sysimg.jl:15
jl_call_fptr_internal at /home/jan/juliadev/src/julia_internal.h:366 [inlined]
jl_call_method_internal at /home/jan/juliadev/src/julia_internal.h:385 [inlined]
jl_apply_generic at /home/jan/juliadev/src/gf.c:2003
process_options at ./client.jl:325
_start at ./client.jl:391
jl_call_fptr_internal at /home/jan/juliadev/src/julia_internal.h:366 [inlined]
jl_call_method_internal at /home/jan/juliadev/src/julia_internal.h:385 [inlined]
jl_apply_generic at /home/jan/juliadev/src/gf.c:2003
jl_apply at /home/jan/juliadev/ui/../src/julia.h:1449 [inlined]
true_main at /home/jan/juliadev/ui/repl.c:107
main at /home/jan/juliadev/ui/repl.c:237
__libc_start_main at /build/glibc-bfm8X4/glibc-2.23/csu/../csu/libc-start.c:291
_start at /home/jan/juliadev/julia (unknown line)
Allocations: 1413922 (Pool: 1413685; Big: 237); GC: 2
Illegal instruction (core dumped)
```

**Comment by jw3126** [comment]
FWIW this is a mvp of a real world code [here](https://github.com/jw3126/ArgCheck.jl/blob/a331be7e2f182f977f8d6e9e70408d9e7e233a89/test/runtests.jl#L124).

**Comment by yuyichao** [comment]
Seems to some issue where macro lowering on 0.7 failed to catch the `$` expression in some cases and then that confuses type inference/codegen...

**Comment by yuyichao** [comment]
More direct repro

```julia
f() = false

macro m2(a, b)
    quote
        while false
        end
        let
            $(esc(b))
        end
    end
end

@m2 "" begin
    global f
    f(m) = false
    if !f(1)
        $QuoteNode
    end
end
```

**Comment by JeffBezanson** [comment]
This is weird, since I thought we had checks for this:

```
julia> @eval f() = $(Expr(:call, Expr(:$, :QuoteNode), 0))
f (generic function with 1 method)

julia> f()
ERROR: error compiling f: syntax: prefix "$" in non-quoted expression

julia> eval(Expr(:$, :x))
ERROR: unsupported or misplaced expression $
```

**Comment by yuyichao** [comment]
Yes, it is strange. The above code would also raise unrelated error if you put it in a function (world age issue on `f`) and won't error at all if the condition is replaced with `false`. It seems that we raise the error too late (should be done in lowering) and the crash probably come from inference assuming an world age while codegen make each global statement run in the latest age.

**Comment by JeffBezanson** [comment]
Repro with no misplaced `$`:

```
f24316() = false

begin
    while false end
    global f24316
    f24316(m) = false
    f24316(1)
end
```

**Comment by yuyichao** [comment]
Yeah, it seems that the missing error for misplaced `$` and the crash are two independent issues......

**Comment by JeffBezanson** [comment]
Fixed by #24600
